Lib/ast.py

```diff
@@ -798,6 +798,9 @@ def traverse(self, node):
         else:
             super().visit(node)
 
+    # Note: as visit() resets the output text, do NOT rely on
+    # NodeVisitor.generic_visit to handle any nodes (as it calls back in to
+    # the subclass visit() method, which resets self._source to an empty list)
```

Comment:
`visit()` as in `_Unparser` is just the top level entry point, methods within this class should use `traverse()` to rely on actual `NodeVisitor.generic_visit`. Though I think this might be something we should change in the future if we ever decide to publicize the API (I get the confusion).

Reply:
Right, the issue is that `NodeVisitor.generic_visit` itself calls `self.visit()` rather than `self.traverse()`, so if `_Unparser` doesn't handle every node you lose all of your previous output. As it turned out, the generic visitor was wrong for `MatchSingleton` anyway, and `MatchValue` is a one-liner, so not relying on it didn't cost me much.